rfcs/2021-07-19-8216-multiple-pipelines.md

```diff
@@ -0,0 +1,261 @@
+This approach would extend the [RFC 2064](https://github.com/timberio/vector/blob/master/rfcs/2020-03-17-2064-event-driven-observability.md#collecting-uniform-context-data) by _just_ adding `pipeline_id` to the context.
+
+In Vector, once [the topology is built from the configuration](https://github.com/timberio/vector/blob/v0.15.0/src/topology/builder.rs#L106), every component is encapsulated in a `Task` that intercepts an incoming event and processes it accordingly. This task also keeps track of its internal metrics and finally emits `internal_metrics` events.
+
+To add the pipeline information to the task, we need to add a new optional parameter to the [`Task::new`](https://github.com/timberio/vector/blob/v0.15.0/src/topology/task.rs#L29) method.
+
+```rust
+pub struct Task {
+    #[pin]
+    inner: BoxFuture<'static, Result<TaskOutput, ()>>,
+    name: String,
+    typetag: String,
+    pipeline: Option<String>,
+}
+impl Task {
+    pub fn new<S1, S2, Fut>(name: S1, typetag: S2, pipeline: Option<String>, inner: Fut) -> Self
+    where
+        S1: Into<String>,
+        S2: Into<String>,
+        Fut: Future<Output = Result<TaskOutput, ()>> + Send + 'static,
+    {
+        Self {
+            inner: inner.boxed(),
+            name: name.into(),
+            typetag: typetag.into(),
+            pipeline,
+        }
+    }
+}
+````
+
+That way, when vector [spawns a new transform task](https://github.com/timberio/vector/blob/v0.15.0/src/topology/mod.rs#L574), it will be able to add the optional pipeline information to the span.
+
+```rust
+let span = error_span!(
+    "transform",
+    component_kind = "transform",
+    component_name = %task.name(),
+    component_type = %task.typetag(),
+    pipeline_id = %task.pipeline_id(),
+);
+```
+
+Doing so, each time the task will emit an internal event, it will be populated by the optional `pipeline_id`.
+
+## Rationale
+
+- Why is this change worth it?
+
+This split improves the readability of the configuration files and allows the users to collaborate, which makes using Vector more user-friendly.
+
+- What is the impact of not doing this?
+
+This would force the users to keep having complex configuration files and/or to duplicate components configuration between their configuration files.
+
+- How does this position us for success in the future?
+
+With this representation, we'll be able add access control by, for example, declaring the pipelines inside the configuration files to limit the reachable components. We would also be able to specify a quota for each pipeline.
+
+## Prior Art
+
+_TODO_
+
+## Drawbacks
+
+- Why should we not do this?
+- What kind of ongoing burden does this place on the team?
+
+## Alternatives
+
+- Do nothing: we can already use several configuration files, people could split their existing configuration.
+
+This would imply some duplication if a transform is used in multiple configuration files.
+Anybody could add a sink or source even if they don't have permission.
```

Comment:
```suggestion
- Do nothing: we can already use several configuration files, people could split their existing configuration. This would imply some duplication if a transform is used in multiple configuration files.
Anybody could add a sink or source even if they don't have permission.
```

Hoisted this second paragraph, to fit within the same bullet point.

It might be worth adding a qualifier to: _"Anybody could add a sink or source even if they don't have permission"_. IMO, this implies Vector is publicly accessible. I take it what you mean here is more that there would be _one_ permission level to whomever has access to the filesystem Vector is reading from vs. being able to separate concerns between a "root" config and a "pipeline".

Reply:
> This would imply some duplication if a transform is used in multiple configuration files.

I guess this is part of the answer but I agree with the rest.